### PR TITLE
Generate Source: Origin Trait Matching

### PR DESCRIPTION
**Pull request by delphiactual**
#611 
https://github.com/DestinyItemManager/DIM/issues/10424
https://github.com/DestinyItemManager/DIM/issues/10425

Might need to remove origin traits for some and add exceptions to others...

### CODE REVIEW

File: output/source-info-v2.ts [hunk 1]
```diff
@@ -992,6 +1045,9 @@ const D2Sources: {
     aliases: ['coil'],
   },
   rootofnightmares: {
+    itemHashes: [
+      1123421440, // Epochal Integration
+    ],
```
Comment (delphiactual) [inline output/source-info-v2.ts:1050]: probably need to remove RoN origin trait matching...

File: output/source-info-v2.ts [hunk 2]
```diff
@@ -1079,23 +1135,66 @@ const D2Sources: {
     aliases: ['sos'],
   },
   spireofthewatcher: {
+    itemHashes: [
+      542573208, // Last Rite
+      959037361, // Wild Style (Adept)
+      1332123064, // Wild Style
+      1968410628, // The Prophet
+      3019024381, // The Prophet (Adept)
+    ],
```
Comment (delphiactual) [inline output/source-info-v2.ts:1144]: and spire of the watcher

File: output/source-info-v2.ts [hunk 3]
```diff
@@ -1079,23 +1135,66 @@ const D2Sources: {
...
+      3001205424, // Ecliptic Distaff
+      3125454907, // Loaded Question
+      3183283212, // Wendigo GL3
       3215252549, // Determination
+      3667553455, // BrayTech Osprey
+      3832743906, // Hung Jury SR4
+      3915197957, // Wendigo GL3 (Adept)
       4060882458, // Balistraria Wrap (Ornament)
+      4074251943, // Hung Jury SR4 (Adept)
+      4162642204, // The Militia's Birthright (Adept)
```
Comment (delphiactual) [inline output/source-info-v2.ts:1197]: add exception to strikes to remove adept

**Review by bhollis — APPROVED** [review]

File: src/generate-source-info.ts [hunk 4]
```diff
@@ -92,32 +92,85 @@ for (const [sourceTag, matchRule] of Object.entries(categories.sources)) {
...
       itemHashes.push(...includedItemHashes);
     }
   }
 
+  if (matchRule.originTrait) {
+    for (const trait of matchRule.originTrait) {
+      const excludedItems = matchRule.excludedItems;
+      const traitHash = allInventoryItems
+        .filter((i) => i.displayProperties.name === trait)
+        .map((i) => i.hash)[0];
```
Comment (delphiactual) [inline src/generate-source-info.ts:143]: this can probably be improved

File: src/generate-source-info.ts [hunk 5]
```diff
@@ -92,32 +92,85 @@ for (const [sourceTag, matchRule] of Object.entries(categories.sources)) {
...
+
+          const traitHashes = [
+            getDef(
+              'PlugSet',
+              i.sockets?.socketEntries.filter((socket) => socket.socketTypeHash === 3993098925)[0]
+                ?.reusablePlugSetHash,
+            ),
+          ]
+            .map((i) => i?.reusablePlugItems.map((p) => p.plugItemHash))
+            .flat();
```
Comment (delphiactual) [inline src/generate-source-info.ts:159]: this also can probably be improved

**Pull request merged**